### PR TITLE
Add BatchPut to storage, add use of context for timeouts

### PR DESCRIPTION
**Pull request by nicpottier**

### CODE REVIEW

**Comment by nicpottier** [comment]
Signature change so we'll need to change mailroom/courier to pass in context when we update.

File: storage/s3.go [hunk 1]
```diff
@@ -102,6 +109,84 @@ func (s *s3Storage) Put(path string, contentType string, contents []byte) (strin
...
+	for w := 0; w < s.workersPerBatch; w++ {
+		go s.batchWorker(ctx, uploads, errors, stop, wg)
+	}
+
+	// add all our uploads to our work queue
+	for _, u := range us {
+		uploads <- u
+	}
+
+	// read all our errors out, we'll stop everything if we encouter one
```
Comment (rowanseymour) [inline storage/s3.go:171]: typo

**Review by rowanseymour — APPROVED** [review]
Excited to watch this in action!

**Comment by nicpottier** [comment]
🍿 🍿 🍿

**Pull request merged**